src/CMakeLists.txt

```diff
@@ -61,6 +61,29 @@ ENDIF ()
 CHECK_FUNCTION_EXISTS(qsort_r HAVE_QSORT_R)
 IF (HAVE_QSORT_R)
 	ADD_DEFINITIONS(-DHAVE_QSORT_R)
+	CHECK_C_SOURCE_COMPILES("
+#include <stdlib.h>
+
+/* BSD */
+void qsort_r(void *base, size_t nmemb, size_t size,
+	void *thunk,
+	int (*compar)(void *, const void *, const void *));
+
+/* glibc */
+/*
+void qsort_r(void *base, size_t nmemb, size_t size,
+	int (*compar)(const void *, const void *, void *),
+	void *arg);
+*/
+
+int main(void)
+{
+	return 0;
+}
+" HAVE_QSORT_R_BSD)
+	IF (HAVE_QSORT_R_BSD)
+		ADD_DEFINITIONS(-DHAVE_QSORT_R_BSD)
+	ENDIF ()
```

Comment:
I think this approach is a lot more sensible. than what we currently have. I'd like to propose a few changes, though:

- instead of defining `HAVE_QSORT_R`, only ever define `HAVE_QSORT_R_BSD` and `HAVE_QSORT_R_GNU` (and adjust locations using the previous define accordingly)
- instead of `CHECK_C_SOURCE_COMPILES` and/or `CHECK_FUNCTION_EXISTS`, we should be using `CHECK_PROTOTYPE_DEFINITION`

So, something like this:

```
CHECK_PROTOTYPE_DEFINITION(qsort_r
	"void qsort_r(void *base, size_t nmemb, size_t size, void *thunk, int (*compar)(void *, const void *, const void *))"
	"NULL" "stdlib.h" HAVE_QSORT_R_BSD)
CHECK_PROTOTYPE_DEFINITION(qsort_r
	"void qsort_r(void *base, size_t nmemb, size_t size, int (*compar)(const void *, const void *, void *), void *arg)"
	"NULL" "stdlib.h" HAVE_QSORT_R_GNU)
```

The only thing is that we'd have to include `CheckPrototypeDefinition`, which is only available since CMake v3.0. There's nothing speaking against distributing it in our cmake/Modules directory, though.